domain/schema/model/sql/0007-charm.sql

```diff
@@ -1,27 +1,398 @@
     PRIMARY KEY (charm_uuid, name)
 );
 
 CREATE INDEX idx_charm_storage_charm
 ON charm_storage (charm_uuid);
+
+CREATE TABLE charm_storage_property (
+    charm_uuid          TEXT NOT NULL,
+    charm_storage_name  TEXT NOT NULL,
+    key                 TEXT,
+    value               TEXT,
+    CONSTRAINT          fk_charm_storage_property_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid),
+    CONSTRAINT          fk_charm_storage_property_charm_storage
+        FOREIGN KEY     (charm_storage_name)
+        REFERENCES      charm_storage(name),
+    PRIMARY KEY (charm_uuid, charm_storage_name, key)
+);
+
+CREATE INDEX idx_charm_storage_property_charm
+ON charm_storage_property (charm_uuid);
+
+CREATE TABLE charm_device (
+    charm_uuid          TEXT NOT NULL,
+    name                TEXT,
+    description         TEXT,
+    device_type         TEXT,
+    count_min           INT NOT NULL,
+    count_max           INT NOT NULL,
+    CONSTRAINT          fk_charm_device_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid),
+    PRIMARY KEY (charm_uuid, name)
+);
+
+CREATE INDEX idx_charm_device_charm
+ON charm_device (charm_uuid);
+
+CREATE TABLE charm_payload (
+    charm_uuid          TEXT NOT NULL,
+    name                TEXT,
+    type                TEXT,
+    CONSTRAINT          fk_charm_payload_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid),
+    PRIMARY KEY (charm_uuid, name)
+);
+
+CREATE INDEX idx_charm_payload_charm
+ON charm_payload (charm_uuid);
+
+CREATE TABLE charm_resource_kind (
+    id       INT PRIMARY KEY,
+    name     TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_charm_resource_kind_name
+    ON charm_resource_kind (name);
+
+INSERT INTO charm_resource_kind VALUES
+    (0, 'file'),
+    (1, 'oci-image');
+
+CREATE TABLE charm_resource (
+    charm_uuid          TEXT NOT NULL,
+    name                TEXT,
+    kind_id             INT NOT NULL,
+    path                TEXT,
+    description         TEXT,
+    CONSTRAINT          fk_charm_resource_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid),
+    CONSTRAINT          fk_charm_resource_charm_resource_kind
+        FOREIGN KEY     (kind_id)
+        REFERENCES      charm_resource_kind(id),
+    PRIMARY KEY (charm_uuid, name)
+);
+
+CREATE INDEX idx_charm_resource_charm
+ON charm_resource (charm_uuid);
+
+CREATE TABLE charm_term (
+    charm_uuid          TEXT NOT NULL,
+    value               TEXT NOT NULL,
+    CONSTRAINT          fk_charm_term_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid),
+    PRIMARY KEY (charm_uuid, value)
+);
+
+CREATE INDEX idx_charm_term_charm
+ON charm_term (charm_uuid);
+
+CREATE TABLE charm_container (
+    charm_uuid          TEXT NOT NULL,
+    name                TEXT,
+    resource            TEXT,
+    uid                 INT,
+    gid                 INT,
```

Comment:
These are optional. And we now model them as *int. So we need a way to reflect that in the schema.